MaxText/train.py

```diff
@@ -267,21 +272,33 @@ def train_step(model, config, state, data, dropout_rng):
     rng2: A new rng key that can be used in future calls.
 
   """
-  train_loss_fn = functools.partial(loss_fn, model, config, data, dropout_rng, is_train=True)
-  grad_fn = jax.value_and_grad(train_loss_fn, has_aux=True)
-  (loss, aux), raw_grads = grad_fn(state.params)
-  intermediate_outputs = aux["intermediate_outputs"]
+  # Initialize grads_accumulation using params
+  grads_accumlation = tree_map(jnp.zeros_like, state.params)
+  raw_grads_accumlation = tree_map(jnp.zeros_like, state.params)
+
+  for grad_accumulate_substep in range(config.gradient_accumulation_steps):
+    train_loss_fn = functools.partial(loss_fn, model, config, data[grad_accumulate_substep], dropout_rng, is_train=True)
+    grad_fn = jax.value_and_grad(train_loss_fn, has_aux=True)
+
+    (loss, aux), raw_grads = grad_fn(state.params)
+    intermediate_outputs = aux["intermediate_outputs"]
+    if config.gradient_clipping_threshold > 0:
+      grads = maxtext_utils.apply_gradient_clipping(raw_grads, state, config.gradient_clipping_threshold)
+    else:
+      grads = raw_grads
 
-  if config.gradient_clipping_threshold > 0:
-    grads = maxtext_utils.apply_gradient_clipping(raw_grads, state, config.gradient_clipping_threshold)
-  else:
-    grads = raw_grads
-  new_state = state.apply_gradients(grads=grads)
+    grads_accumlation = tree_map(lambda x, y: x + y, grads_accumlation, grads)
+    raw_grads_accumlation = tree_map(lambda x, y: x + y, raw_grads_accumlation, raw_grads)
+
+  final_grads = tree_map(lambda x: x / config.gradient_accumulation_steps, grads_accumlation)
```

Comment:
For 'FP8' we have these special 'overwrite_with_gradient' collections with quantization scales/histories. We shouldn't sum those 'gradients' - just using the latest one should be the right thing I think

Reply:
Do you mean `FP8` for GPU?

So maybe we should skip this `gradient_accumulation_steps` feature when `dtype` is `fp8`?

cc: @michelle-yooh

Reply:
Yes I mean when `quantization=fp8`. And thinking more about this again, I don't think using the last one like I mentioned above would work super great - lot of history being lost. 

I think for now lets just skip/assert both can't be used. The new version of FP8 I'm working on has a different approach for updating these scales/history that I think will work better with this, so we can revisit then.

Reply:
Wondering if not just 'FP8' to use the last grad. Why not all use the last grad?

Reply:
Added the assert to make sure `quantization=fp8` can't be used with `gradient_accumulation_steps`